### PR TITLE
KAA-1367: Failed build and linking C++ part on OS X

### PR DESCRIPTION
**Pull request by vasner**
Fix compiler warnings that are generated while tests compilation;
Fix static casting error;
Compilation errors fixes;

Reviewers: @forGGe and @rasendubi

### CODE REVIEW

File: client/client-multi/client-cpp/kaa/configuration/ConfigurationTransport.hpp [hunk 1]
```diff
@@ -50,6 +50,8 @@ class ConfigurationTransport : public AbstractKaaTransport<TransportType::CONFIG
         configurationProcessor_ = processor;
     }
 
+    virtual ~ConfigurationTransport() noexcept {};
```
Comment (rasendubi) [inline client/client-multi/client-cpp/kaa/configuration/ConfigurationTransport.hpp]: Trailing `;` is not needed

File: client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp [hunk 2]
```diff
@@ -70,6 +70,8 @@ class ConfigurationManager : public IConfigurationManager,
         return configurationHash_;
     }
 
+    virtual ~ConfigurationManager() noexcept {};
```
Comment (rasendubi) [inline client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp]: Trailing `;` is not needed

**Comment by rasendubi** [comment]
Minor nitpicks. Otherwise, looks good to me

File: client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp (hunk 2 above)
Comment (vasner) [inline client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp]: Fixed.

Comment (rasendubi) [inline client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp]: it is not

Comment (vasner) [inline client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp]: :) Forgot second place. Done.

**Comment by rasendubi** [comment]
LGTM

**Comment by vadimol** [comment]
+

File: client/client-multi/client-cpp/CMakeLists.txt [hunk 3]
```diff
@@ -204,6 +204,10 @@ if(NOT MSVC)
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -ggdb")
 endif()
 
+# "no-unused-private-field" is added since in most of cases it is false positive
+list(APPEND KAA_COMPILE_OPTIONS
+                -Wno-unused-private-field)
```
Comment (ghost) [inline client/client-multi/client-cpp/CMakeLists.txt:209]: Does the flag exist in MSVC ?

Comment (ghost) [inline client/client-multi/client-cpp/CMakeLists.txt:209]: And what fields are unused?

Comment (vasner) [inline client/client-multi/client-cpp/CMakeLists.txt:209]: > Does the flag exist in MSVC ?

In MSVC exists "Unused private fields (CA1823)" not sure "-Wno-unused-private-field" works.
I will appreciate, if you can test. I can add "if else" after.

> And what fields are unused?

While compilation of Events C++ demo on OS X with gcc I got lot of these dummy warnings:
In file included from /Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/impl/Kaa.cpp:21:
In file included from /Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/test/../kaa/KaaClient.hpp:28:
/Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/test/../kaa/profile/ProfileManager.hpp:107:39: warning: 
      private field 'context_' is not used [-Wunused-private-field]
    IKaaClientContext                &context_;

This field is used (if I comment private filed this module will not compile), but warning is exists. We discussed this with @rasendubi and decided to add this warning stopper.

File: client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp [hunk 4]
```diff
@@ -70,6 +70,8 @@ class ConfigurationManager : public IConfigurationManager,
         return configurationHash_;
     }
 
+    virtual ~ConfigurationManager() noexcept {}
```
Comment (ghost) [inline client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp:73]: Could you explain why we need the desctuctor?

File: client/client-multi/client-cpp/kaa/event/EventManager.hpp [hunk 5]
```diff
@@ -70,13 +70,15 @@ class EventManager : public IEventManager
 
     virtual void setTransport(EventTransport *transport);
 
+    using AbstractTransactable::beginTransaction;
```
Comment (ghost) [inline client/client-multi/client-cpp/kaa/event/EventManager.hpp:73]: As I understand, without this statement code doesn't compile?

Comment (vasner) [inline client/client-multi/client-cpp/kaa/event/EventManager.hpp:73]: No, this is just warning fix:
In file included from /Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/impl/KaaClient.cpp:19:
In file included from /Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/test/../kaa/KaaClient.hpp:24:
/Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/test/../kaa/event/EventManager.hpp:73:30: warning: 
      'kaa::EventManager::beginTransaction' hides overloaded virtual function [-Woverloaded-virtual]
    virtual TransactionIdPtr beginTransaction()
                             ^
/Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/test/../kaa/transact/AbstractTransactable.hpp:30:30: note: 
      hidden overloaded virtual function 'kaa::AbstractTransactable<std::__1::list<kaa::Event, std::__1::allocator<kaa::Event> > >::beginTransaction' declared
      here: different number of parameters (1 vs 0)
    virtual TransactionIdPtr beginTransaction(IKaaClientContext &context_) {
                             ^

File: client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp (hunk 4 above)
Comment (vasner) [inline client/client-multi/client-cpp/kaa/configuration/manager/ConfigurationManager.hpp:73]: I got this error:
In file included from /Users/vasiliy/Works/Temp/2016.08.16/CppEventDemo/libs/kaa/kaa/KaaClient.hpp:37:0,
                 from /Users/vasiliy/Works/Temp/2016.08.16/CppEventDemo/libs/kaa/impl/Kaa.cpp:21:
/Users/vasiliy/Works/Temp/2016.08.16/CppEventDemo/libs/kaa/kaa/configuration/manager/ConfigurationManager.hpp:42:7: error: looser throw specifier for 'virtual kaa::ConfigurationManager::~ConfigurationManager()'
In file included from /Users/vasiliy/Works/Temp/2016.08.16/CppEventDemo/libs/kaa/kaa/configuration/manager/ConfigurationManager.hpp:24:0,
                 from /Users/vasiliy/Works/Temp/2016.08.16/CppEventDemo/libs/kaa/kaa/KaaClient.hpp:37,
                 from /Users/vasiliy/Works/Temp/2016.08.16/CppEventDemo/libs/kaa/impl/Kaa.cpp:21:
/Users/vasiliy/Works/Temp/2016.08.16/CppEventDemo/libs/kaa/kaa/configuration/manager/IConfigurationManager.hpp:72:13: error:   overriding 'virtual kaa::IConfigurationManager::~IConfigurationManager() noexcept (true)'

File: client/client-multi/client-cpp/CMakeLists.txt (hunk 3 above)
Comment (ghost) [inline client/client-multi/client-cpp/CMakeLists.txt:209]: Well, I had removed any mention of IKaaClientContext from ProfileManager.hpp and had edited KaaClient.cpp and SDK successfully compiled. Did I understand you correctly?

Comment (vasner) [inline client/client-multi/client-cpp/CMakeLists.txt:209]: I do not "remove any mention of IKaaClientContext from ProfileManager.hpp and had edited KaaClient.cpp". I just added flag to do not see this warning any more. 

I do not see mentions of "IKaaClientContext and ProfileManager.hpp" while compiling this module:
Building CXX object CMakeFiles/kaatest.dir/Users/vasiliy/Works/CyberVision/Internal/kaa/git-repos/kaa-vnerozin/client/client-multi/client-cpp/impl/logging/DefaultLogger.cpp.o

Comment (ghost) [inline client/client-multi/client-cpp/CMakeLists.txt:209]: I mean if it possible maybe we can get rid of suppressing the unused-private-members warning and fix the reason.

Comment (vasner) [inline client/client-multi/client-cpp/CMakeLists.txt:209]: I understood. I tried to tell that I do not see the reason :) I cannot find any place in DefaultLogger.cpp to put usage of private field 'context_' from ProfileManager.hpp

**Pull request merged**